src/jupyterlabcontroller/services/builder/lab.py

```diff
@@ -0,0 +1,821 @@
+                    )
+                case PVCVolumeSource():
+                    claim = V1PersistentVolumeClaimVolumeSource(
+                        claim_name=f"{username}-nb-pvc-{pvc_count}",
+                        read_only=is_readonly,
+                    )
+                    volume = V1Volume(name=name, persistent_volume_claim=claim)
+                    pvc_count += 1
+            mount = V1VolumeMount(
+                name=name,
+                mount_path=prefix + storage.container_path,
+                sub_path=storage.sub_path,
+                read_only=is_readonly,
+            )
+            container = LabVolumeContainer(volume=volume, volume_mount=mount)
+            results.append(container)
+        return results
+
+    def namespace_for_user(self, username: str) -> str:
+        """Construct the namespace name for a user's lab environment.
+
+        Parameters
+        ----------
+        username
+            Username of lab user.
+
+        Returns
+        -------
+        str
+            Name of their namespace.
+        """
+        return f"{self._config.namespace_prefix}-{username}"
+
+    def recreate_env(self, env: dict[str, str]) -> dict[str, str]:
+        """Recreate the JupyterHub-provided environment.
+
+        When reconciling state from Kubernetes, we need to recover the content
+        of the environment sent from JupyterHub from the ``ConfigMap`` in the
+        user's lab environment. We can't recover the exact original
+        environment, but we can recreate an equivalent one by filtering out
+        the environment variables that would be set directly by the lab
+        controller.
+
+        Parameters
+        ----------
+        env
+            Environment recovered from a ``ConfigMap``.
+
+        Returns
+        -------
+        dict of str
+            Equivalent environment sent by JupyterHub.
+
+        Notes
+        -----
+        The list of environment variables that are always added internally by
+        the lab controller must be kept in sync with the code that creates the
+        config map.
+        """
+        unwanted = {
+            "CPU_GUARANTEE",
+            "CPU_LIMIT",
+            "DEBUG",
+            "EXTERNAL_INSTANCE_URL",
+            "IMAGE_DESCRIPTION",
+            "IMAGE_DIGEST",
+            "JUPYTER_IMAGE",
+            "JUPYTER_IMAGE_SPEC",
+            "MEM_GUARANTEE",
+            "MEM_LIMIT",
+            "RESET_USER_ENV",
+            *list(self._config.env.keys()),
+        }
+        return {k: v for k, v in env.items() if k not in unwanted}
+
+    def _build_home_directory(self, username: str) -> str:
+        """Construct the home directory path for a user."""
+        match self._config.homedir_schema:
+            case UserHomeDirectorySchema.USERNAME:
+                return f"/home/{username}"
+            case UserHomeDirectorySchema.INITIAL_THEN_USERNAME:
+                return f"/home/{username[0]}/{username}"
```

Comment:
Should we just go ahead and parameterize /home while we're in here, since we're about to need it?

Reply:
That's a logically separate change, so I'd rather do it as a separate PR.